frontend/src/pages/PipelineDetails.test.tsx

```diff
@@ -0,0 +1,328 @@
+import { graphlib } from 'dagre';
+import { shallow, mount } from 'enzyme';
+
+describe('PipelineDetails', () => {
+  const updateBannerSpy = jest.fn();
+  const updateDialogSpy = jest.fn();
+  const updateSnackbarSpy = jest.fn();
+  const updateToolbarSpy = jest.fn();
+  const historyPushSpy = jest.fn();
+  const getPipelineSpy = jest.spyOn(Apis.pipelineServiceApi, 'getPipeline');
+  const deletePipelineSpy = jest.spyOn(Apis.pipelineServiceApi, 'deletePipeline');
+  const getTemplateSpy = jest.spyOn(Apis.pipelineServiceApi, 'getTemplate');
+  const createGraphSpy = jest.spyOn(StaticGraphParser, 'createGraph');
+
+  let testPipeline: ApiPipeline = {};
+
+  function generateProps(): PageProps {
+    // getInitialToolbarState relies on page props having been populated, so fill those first
+    const pageProps: PageProps = {
+      history: { push: historyPushSpy } as any,
+      location: '' as any,
+      match: { params: { [RouteParams.pipelineId]: testPipeline.id }, isExact: true, path: '', url: '' },
+      toolbarProps: { actions: [], breadcrumbs: [], pageTitle: '' },
+      updateBanner: updateBannerSpy,
+      updateDialog: updateDialogSpy,
+      updateSnackbar: updateSnackbarSpy,
+      updateToolbar: updateToolbarSpy,
+    };
+    return Object.assign(pageProps, {
+      toolbarProps: new PipelineDetails(pageProps).getInitialToolbarState(),
+    });
+  }
+
+  beforeAll(() => jest.spyOn(console, 'error').mockImplementation());
+  afterAll(() => jest.resetAllMocks());
+
+  beforeEach(() => {
+    testPipeline = {
+      created_at: new Date(2018, 8, 5, 4, 3, 2),
+      description: 'test job description',
+      enabled: true,
+      id: 'test-job-id',
+      max_concurrency: '50',
+      name: 'test job',
+      pipeline_spec: {
+        parameters: [{ name: 'param1', value: 'value1' }],
+        pipeline_id: 'some-pipeline-id',
+      },
+      trigger: {
+        periodic_schedule: {
+          end_time: new Date(2018, 10, 9, 8, 7, 6),
+          interval_second: '3600',
+          start_time: new Date(2018, 9, 8, 7, 6),
+        }
+      },
+    } as ApiPipeline;
+
+    historyPushSpy.mockClear();
+    updateBannerSpy.mockClear();
+    updateDialogSpy.mockClear();
+    updateSnackbarSpy.mockClear();
+    updateToolbarSpy.mockClear();
+    deletePipelineSpy.mockReset();
+    getPipelineSpy.mockImplementation(() => Promise.resolve(testPipeline));
+    getPipelineSpy.mockClear();
+    getTemplateSpy.mockImplementation(() => Promise.resolve('{}'));
+    getTemplateSpy.mockClear();
+    createGraphSpy.mockImplementation(() => new graphlib.Graph());
+    createGraphSpy.mockClear();
+  });
+
+  it('shows empty pipeline details with no graph graph', async () => {
+    TestUtils.makeErrorResponseOnce(createGraphSpy, 'bad graph');
+    const tree = shallow(<PipelineDetails {...generateProps()} />);
+    await getTemplateSpy;
+    await TestUtils.flushPromises();
+    expect(tree).toMatchSnapshot();
+    tree.unmount();
```

Comment:
why not just call `unmount` in an `afterEach` ?

Reply:
Fair. Done.